src/boundary/boundary_p.cc

```diff
@@ -823,4 +1015,166 @@ boundary_p( particle_bc_t       * RESTRICT pbc_list,
+         printf("%d ", *print_cab_int);
+         print_cab_int++;
+         j += sizeof(int);
+
+         // Print ID if applicable
+         #ifdef VPIC_GLOBAL_PARTICLE_ID
+         size_t* print_cab_sizet = (size_t*) (print_cab +j);
+         printf("%ld", *print_cab_sizet);
+         print_cab_sizet++;
+         j += sizeof(size_t);
+         #endif
+
+         //Print annotation
+         for(; j<max_cas; j+=sizeof(float)) {
+           float* print_cab_float = (float*) (print_cab + j);
+           printf(" %f", *print_cab_float);
+         }
+
+         printf(", ");
+       }
+       printf("]\n");
+      }
+    }
+  } */
+
+  // Ensure sufficent buffer size and prepost recv
+  for( face = 0; face < 6; face++ ) {
+    if ( shared[ face ] ) {
+      if(n_recv[face] > 0) {
+        mp_size_recv_buffer( mp, f2b[ face ], max_cas * n_recv[face] );
+
+        mp_begin_recv( mp, f2b[ face ], max_cas * n_recv[face], bc[ face ], f2rb[ face ] );
+      }
+      if(n_send[face] > 0) {
+        mp_size_send_buffer( mp, f2b[ face ], max_cas * n_send[face] );
+      }
+    }
+  }
+
+  // Fill and send buffers
+  for( face = 0; face < 6; face++ ) {
+    if ( shared[ face ] ) {
+      if(n_send[face] > 0) {
+        float* send_buf = (float *) mp_send_buffer( mp, f2b[ face ] );
+
+        memcpy(send_buf, cab[face], max_cas*n_send[face]);
+
+        mp_begin_send( mp, f2b[ face ], max_cas*n_send[face], bc[ face ], f2b[ face ] );
+      }
+    }
+  }
+
+  // Wait for recv and handle buffer
+  for( face = 0; face < 6; face++ ) {
+    if ( shared[ face ] ) {
+      if(n_recv[face] > 0) {
+        mp_end_recv( mp, f2b[face] );
+
+        /* if(rank() == 0) {
+          // Print for debugging purposes
+          printf("<%d> recv_buffer[%d] = [", rank(), face);
+          for(int i = 0; i<n_recv[face]; i++) {
+            char* recv_buffer = (char*) mp_recv_buffer(mp,f2b[face]);
+            // Begininng of particle
+            char* print_cab = &(recv_buffer[i*max_cas]);
+            int j = 0;
+
+            //Print species ID
+            int* print_cab_int = (int*) print_cab;
+            printf("%d ", *print_cab_int);
+            print_cab_int++;
+            j += sizeof(int);
+
+            // Print ID if applicable
+            #ifdef VPIC_GLOBAL_PARTICLE_ID
+            size_t* print_cab_sizet = (size_t*) (print_cab +j);
+            printf("%ld", *print_cab_sizet);
+            print_cab_sizet++;
+            j += sizeof(size_t);
+            #endif
+
+            //Print annotation
+            for(; j<max_cas; j+=sizeof(float)) {
+              float* print_cab_float = (float*) (print_cab + j);
+              printf(" %f", *print_cab_float);
+            }
+
+            printf(", ");
+          }
+          printf("]\n");
+        } */
+
+        // Unpack buffer and store attributes
+        for(int i = n_recv[face]-1; i>=0; i--) {
+          char* recv_buffer = ((char*) mp_recv_buffer(mp,f2b[face])) + i*max_cas ;
```

Comment:
This stuff doesn't look so bad once someone else has gone through the pain of figuring it out! Nice job

Reply:
once you actually understand what the original code does it's not that bad.